autopilot/scanner/scanner.go

```diff
@@ -0,0 +1,306 @@
+package scanner
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/stats"
+	"go.uber.org/zap"
+)
+
+const (
+	DefaultScanTimeout = 10 * time.Second
+)
+
+type (
+	HostScanner interface {
+		RHPScan(ctx context.Context, hostKey types.PublicKey, hostIP string, timeout time.Duration) (api.RHPScanResponse, error)
+	}
+
+	HostStore interface {
+		HostsForScanning(ctx context.Context, opts api.HostsForScanningOptions) ([]api.HostAddress, error)
+		RemoveOfflineHosts(ctx context.Context, minRecentScanFailures uint64, maxDowntime time.Duration) (uint64, error)
+	}
+
+	Scanner interface {
+		Scan(ctx context.Context, w HostScanner, force bool)
+		Shutdown(ctx context.Context) error
+		Status() (bool, time.Time)
+		UpdateHostsConfig(cfg api.HostsConfig)
+	}
+)
+
+type (
+	scanner struct {
+		hs HostStore
+
+		scanBatchSize int
+		scanThreads   int
+		scanInterval  time.Duration
+
+		statsHostPingMS *stats.DataPoints
+
+		shutdownChan chan struct{}
+		wg           sync.WaitGroup
+
+		logger *zap.SugaredLogger
+
+		mu       sync.Mutex
+		hostsCfg *api.HostsConfig
+
+		scanning          bool
+		scanningLastStart time.Time
+
+		interruptChan chan struct{}
+	}
+
+	scanJob struct {
+		hostKey types.PublicKey
+		hostIP  string
+	}
+)
+
+func New(hs HostStore, scanBatchSize, scanThreads uint64, scanMinInterval time.Duration, logger *zap.SugaredLogger) (Scanner, error) {
+	if scanBatchSize == 0 {
+		return nil, errors.New("scanner batch size has to be greater than zero")
+	}
+	if scanThreads == 0 {
+		return nil, errors.New("scanner threads has to be greater than zero")
+	}
+	return &scanner{
+		hs: hs,
+
+		scanBatchSize: int(scanBatchSize),
+		scanThreads:   int(scanThreads),
+		scanInterval:  scanMinInterval,
+
+		statsHostPingMS: stats.NoDecay(),
+		logger:          logger.Named("scanner"),
+
+		interruptChan: make(chan struct{}),
+		shutdownChan:  make(chan struct{}),
+	}, nil
+}
+
+func (s *scanner) Scan(ctx context.Context, w HostScanner, force bool) {
```

Comment:
Yeah it's odd, I kept the `w` to indicate it's just a worker. We pass the worker because we randomly select a worker from the pool to do the scanning, I think we should keep that because at one point the worker pool will be more dynamic. I had a hard time coming up with a good name, I renamed it now to `WorkerRHPScan` which doesn't roll off the tongue but in those cases it's probably best to just specify exactly what it is. Wdyt?